data/src/main/java/com/microsoft/azure/kusto/data/auth/CloudDependantTokenProviderBase.java

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package com.microsoft.azure.kusto.data.auth;
+
+import com.microsoft.azure.kusto.data.exceptions.DataClientException;
+import com.microsoft.azure.kusto.data.exceptions.DataServiceException;
+import java.net.URISyntaxException;
+import java.util.HashSet;
+import java.util.Set;
+import org.jetbrains.annotations.NotNull;
+
+public abstract class CloudDependantTokenProviderBase extends TokenProviderBase {
+    private static final String ERROR_INVALID_SERVICE_RESOURCE_URL =
+            "Error determining scope due to invalid Kusto Service Resource URL";
+    protected final Set<String> scopes = new HashSet<>();
+    protected CloudInfo cloudInfo = null;
+
+    CloudDependantTokenProviderBase(@NotNull String clusterUrl) throws URISyntaxException {
+        super(clusterUrl);
+    }
+
+    protected void onCloudInfoInitialized() throws DataClientException, DataServiceException {
+        try {
+            scopes.add(cloudInfo.determineScope());
+        } catch (URISyntaxException e) {
+            throw new DataServiceException(clusterUrl, ERROR_INVALID_SERVICE_RESOURCE_URL, e, true);
+        }
+    }
+
+    protected void initializeCloudInfo() throws DataClientException, DataServiceException {
+        if (cloudInfo != null) {
+            return;
+        }
+        synchronized (this) {
+            if (cloudInfo != null) {
```

Comment:
We might've discussed this before, but I don't think double-checked locking works in Java without ensuring JRE optimizations didn't partially initialize `cloudInfo`.

Reply:
I removed the double-checking, now it's one synchronized method. WDYT?

Reply:
Looks great! Your quick escape if not null should ensure solid performance.